### DISCUSSION TITLE
How would you go by handling errors like these?

### DISCUSSION THREAD

**Discussion by lorikku** [vercel/next.js#61131] [Help]
### Summary

Today I got the following error in my application:

![image](https://github.com/vercel/next.js/assets/56921718/a37df8af-1676-4254-8d7c-e6bb1b4708cb)

I know why and where this error occurred, because I'm actively working on this project and it was an error expected to happen. But I just kept asking myself, what would I do if I'm new to this project and got this error?

I would have no clue where the error had happened (which component, function, etc.) Is this the result of poor error handling? How can I make sure I handle stuff like this properly?

Just FYI: this project uses TS with properly set up types, it was merely a change to our CMS structure that caused this error (obviously), but a scenario I could think of is eg. our CMS provider makes a change to their system (without letting us know beforehand) -> errors like these happen, etc. I think you should be able to locate the error in any context, no matter what caused it.

**Comment by icyJoseph** [comment]
> this project uses TS with properly set up types, it was merely a change to our CMS structure that caused this error

So what used to be a string is now an object? Since that's a proven possibility, given the error you see, you ought to either:

- Create Types that more closely reflect what the CMS can do, and not just what it currently does

Think of a `type Article = { title: string; }`, but in reality the CMS also supports multi-lang titles, as an object. Then your types should truly reflect that, so that if an object is found, you have code in place to handle that case.

- Validate the data coming from the CMS

The previous point breaks down when the CMS actually does change what it can do. Then it is good to have the mindset that, external data can't really be trusted.

In this case, if your client side can afford it, you should validate the CMS data, either through superstruct, zod, custom made type assertions, etc...

Think of `await res.json()`, that one returns `any`, and many people straight up cast that to an expected type shape, but you should probably, at least, make some assertions on that data, because, again, it is external and you can't trust its shape.

People who use Next.js as a Backend for the frontend, sometimes, use an API route to the heavy lifting of parsing and validating. Still, they then trust that the API route returns the correctly shaped data to the client.

- Error boundaries

Use these around areas that handle data that you don't trust, so that if, while rendering, an error happens, you can at least get more information, and show a fallback UI to the affected React sub tree.

Regarding finding where this happens, I think that the Stack contains more info about where it happened right? If you expand the frames that is.

**Reply by lorikku** [reply]
Thanks so much for the tips! "Validate the data coming from the CMS" -> lesson learned that external data cannot be trusted indeed!

Concerning your comment about:
> Regarding finding where this happens, I think that the Stack contains more info about where it happened right? If you expand the frames that is.

The stack does not actually show me anything useful of where the error happened. Expanding the frames, I just get the exact same stack trace which was already logged in the error message:

![image](https://github.com/vercel/next.js/assets/56921718/9f0982b4-a896-4a6d-97cc-55dd024e28de)

So here I'm still left a bit confused why there's not even a component mentioned anywhere of where the error happened. Not a useful stack trace. Can this somehow be 'improved'?

**Reply by icyJoseph** [reply]
Does the Error Overlay show something else? 

Faced with this, I'd put a Chrome Dev Tools debugger on `throwOnInvalidObjectType`, and step through, most likely I'd find the issue there.

**Reply by lorikku** [reply]
Error Overlay shows the same:

![image](https://github.com/vercel/next.js/assets/56921718/89a4cd86-5b09-4c09-bb44-1f713ebcfc36)

It's a server component, so I can't use Chrome Dev Tools 😅

**Comment by lorikku** [comment]
Seems like I'm not the only one having issues with stack traces
- https://github.com/vercel/next.js/discussions/55705
- https://github.com/vercel/next.js/issues/4613
